plugin/hijack.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"bufio"
+	"errors"
+	"net"
+	"net/http"
+	"net/rpc"
+	"time"
+)
+
+const (
+	hijackedConnReadBufSize = 4096
+)
+
+var (
+	ErrNotHijacked     = errors.New("response is not hijacked")
+	ErrAlreadyHijacked = errors.New("response was already hijacked")
+	ErrCannotHijack    = errors.New("response cannot be hijacked")
+)
+
+func (w *httpResponseWriterRPCServer) HjConnRWRead(b []byte, reply *[]byte) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	data := make([]byte, len(b))
+	n, err := w.hjr.bufrw.Read(data)
+	if err != nil {
+		return err
+	}
+	*reply = data[:n]
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnRWWrite(b []byte, reply *int) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	n, err := w.hjr.bufrw.Write(b)
+	if err != nil {
+		return err
+	}
+	*reply = n
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnRead(size int, reply *[]byte) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	if len(w.hjr.readBuf) < size {
+		w.hjr.readBuf = make([]byte, size)
+	}
+	n, err := w.hjr.conn.Read(w.hjr.readBuf[:size])
+	if err != nil {
+		return err
+	}
+	*reply = w.hjr.readBuf[:n]
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnWrite(b []byte, reply *int) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	n, err := w.hjr.conn.Write(b)
+	if err != nil {
+		return err
+	}
+	*reply = n
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnClose(args struct{}, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.Close()
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetReadDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetReadDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetWriteDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetWriteDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HijackResponse(args struct{}, reply *struct{}) error {
+	if w.hjr != nil {
+		return ErrAlreadyHijacked
+	}
+	hj, ok := w.w.(http.Hijacker)
+	if !ok {
+		return ErrCannotHijack
+	}
+	conn, bufrw, err := hj.Hijack()
+	if err != nil {
+		return err
+	}
+
+	w.hjr = &hijackedResponse{
+		conn:    conn,
+		bufrw:   bufrw,
+		readBuf: make([]byte, hijackedConnReadBufSize),
+	}
+	return nil
+}
```

Comment:
Why is this method needed? It doesn't seam to be called anywhere in the code nor is it needed to fulfill an interface.

Reply:
It's not called directly but through RPC:

https://github.com/mattermost/mattermost-server/blob/cae976643196afd812e106aa477025141fae49ba/plugin/hijack.go#L200-L202

Reply:
Got it, thanks for the explanation :+1: